packages/mdc-data-table/foundation.ts

```diff
@@ -0,0 +1,193 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+import {MDCFoundation} from '@material/base/foundation';
+import {MDCDataTableAdapter} from './adapter';
+import {cssClasses, strings} from './constants';
+
+export class MDCDataTableFoundation extends MDCFoundation<MDCDataTableAdapter> {
+  static get defaultAdapter(): MDCDataTableAdapter {
+    return {
+      addClassAtRowIndex: () => undefined,
+      getRowCount: () => 0,
+      getRowElements: () => [],
+      getRowIdAtIndex: () => '',
+      getRowIndexByChildElement: () => 0,
+      getSelectedRowCount: () => 0,
+      isCheckboxAtRowIndexChecked: () => false,
+      isHeaderRowCheckboxChecked: () => false,
+      isRowsSelectable: () => false,
+      notifyRowSelectionChanged: () => undefined,
+      notifySelectedAll: () => undefined,
+      notifyUnselectedAll: () => undefined,
+      registerHeaderRowCheckbox: () => undefined,
+      registerRowCheckboxes: () => undefined,
+      removeClassAtRowIndex: () => undefined,
+      setAttributeAtRowIndex: () => undefined,
+      setHeaderRowCheckboxChecked: () => undefined,
+      setHeaderRowCheckboxIndeterminate: () => undefined,
+      setRowCheckboxCheckedAtIndex: () => undefined,
+    };
+  }
+
+  constructor(adapter?: Partial<MDCDataTableAdapter>) {
+    super({...MDCDataTableFoundation.defaultAdapter, ...adapter});
+  }
+
+  init() {
+    this.layout();
+  }
+
+  /**
+   * Re-initializes header row checkbox and row checkboxes when selectable rows are added or removed from table.
+   * Use this if registering checkbox is synchronous.
+   */
+  layout() {
+    if (this.adapter_.isRowsSelectable()) {
+      this.adapter_.registerHeaderRowCheckbox();
+      this.adapter_.registerRowCheckboxes();
+
+      this.setHeaderRowCheckboxState_();
+    }
+  }
+
+  /**
+   * Re-initializes header row checkbox and row checkboxes when selectable rows are added or removed from table.
+   * Use this if registering checkbox is asynchronous.
+   */
+  async layoutAsync(): Promise<void> {
+    if (this.adapter_.isRowsSelectable()) {
+      await this.adapter_.registerHeaderRowCheckbox();
+      await this.adapter_.registerRowCheckboxes();
+
+      this.setHeaderRowCheckboxState_();
+    }
+  }
+
+  /**
+   * @return Returns array of row elements.
+   */
+  getRows(): Element[] {
+    return this.adapter_.getRowElements();
+  }
+
+  /**
+   * Sets selected row ids. Overwrites previously selected rows.
+   * @param rowIds Array of row ids that needs to be selected.
+   */
+  setSelectedRowIds(rowIds: string[]) {
+    for (let rowIndex = 0; rowIndex < this.adapter_.getRowCount(); rowIndex++) {
+      const rowId = this.adapter_.getRowIdAtIndex(rowIndex);
+
+      let isSelected = false;
```

Comment:
you could condense this to
```
const isSelected = rowId && rowIds.indexOf(rowId) >= 0;
```

Reply:
I didn't want to use ternary operator since the condition is complex.